blog/_posts/2018-02-07-collections-performance.md

```diff
@@ -0,0 +1,136 @@
+---
+layout: blog-detail
+post-type: blog
+by: Julien Richard-Foy
+title: On Performance of the New Collections
+---
+
+In a [previous blog post](/blog/2017/11/28/view-based-collections.html), I explained
+how [Scala 2.13’s new collections](http://www.scala-lang.org/blog/2017/02/28/collections-rework.html)
+have been designed so that the default implementations of transformation operations work
+with both strict and non-strict types of collections. In essence, we abstract over
+the evaluation mode (strict or non strict) of concrete collection types.
+
+After we published that blog post, the community
+[raised concerns](https://www.reddit.com/r/scala/comments/7g52cy/let_them_be_lazy/dqgol36/)
+about possible performance implications of having more levels of abstraction than before.
+
+This blog article gives more information about the overhead of the
+collections’ view-based design and our solution to remove that
+overhead.
+
+For reference, the source code of the new collections is available in
+[this GitHub repository](https://github.com/scala/collection-strawman).
+
+## Overhead Of View Based Implementations
+
+Let’s be clear: the view based implementations are in general slower than their
+builder based versions. How much slower exactly varies with the type of collection
+(e.g. `List`, `Vector`, `Set`), the operation (e.g. `map`, `flatMap`, `filter`)
+and the number of elements in the collection. In my benchmark on `Vector`, on
+the `map`, `filter` and `flatMap` operations, with 1 element to 7 million of
+elements, I measured an average slowdown of 25%.
+
+## How To Fix That Performance Regression?
+
+Our solution is simply to go back to builder based implementations for strict collections: we
+override the default view based implementations with more efficient builder based
+ones. We actually and up with the same implementations as in the old collections.
+
+In practice these implementations are factored out in traits that can be mixed
+into concrete collection types. Such trait names are always prefixed with
+`StrictOptimized`. For instance, here is an excerpt of the `StrictOptimizedIterableOps`
+trait:
+
+~~~ scala
+trait StrictOptimizedIterableOps[+A, +CC[_], +C] extends IterableOps[A, CC, C] {
+
+  override def map[B](f: A => B): CC[B] = {
+    val b = iterableFactory.newBuilder[B]()
+    val it = iterator()
+    while (it.hasNext) {
+      b += f(it.next())
+    }
+    b.result()
+  }
+
+}
+~~~
+
+Then, to implement the `Vector` collection, we just mix the above trait:
+
+~~~ scala
+trait Vector[+A] extends IndexedSeq[A]
+  with IndexedSeqOps[A, Vector, Vector[A]]
+  with StrictOptimizedSeqOps[A, Vector, Vector[A]]
+~~~
+
+Here we use `StrictOptimizedSeqOps`, which is a specialization of `StrictOptimizedIterableOps`
+for `Seq` collections.
+
+## Is The View Based Design Worth It?
+
+In my previous article I explained that a drawback of the old builder based design was that,
```

Comment:
This sentence could be split into two.

In my previous article, I explained that a drawback of the old builder based design. On non-strict collections (e.g. Stream or View), we had to carefully override all the default implementations of transformation operations to make them non-strict.